### DISCUSSION TITLE
TLS MQTT - Auto host discovery with self signed cert (build flag exclusion conflict)

### DISCUSSION THREAD

**Discussion by nccchirag** [arendst/Tasmota#16814] [Q&A] [answered]
I followed the [guide](https://tasmota.github.io/docs/Self-signed-Mosquitto/)  and I'm able to securely establish the connection as intended. 

My question is that how I can use Auto Discovery? Because during the build I get mutual exclusion error -  _#error "Select either USE_DISCOVERY or USE_MQTT_AWS_IOT, mDNS takes too much ..._

with auto discovery flags undefined the build compiles and I can establish secure connection to the server, but I need to manually configure the host IP in MQTT Config section!

The command TLSKey does not work without the flag USE_MQTT_AWS_IOT
![image](https://user-images.githubusercontent.com/7569367/195848800-cb0706ae-cfd3-4a03-8204-fab51b405561.png)

Is there anyway other by which I can achieve auto discovery between tasmota client & Mosquitto server?

**Comment by sfromis** [comment]
Well, nothing stops you from removing that check from `my_user_config.h`, and see if you can shoehorn it in to also work. Maybe it takes giving up the option of OTA updating to newer builds. Obviously the check was not added without reason, so....

Anyway, I'm "unconvinced" that the Tasmota discovery payloads intended to work with the Tasmota integration in HA is helpful to other types of configuration wants.

**Reply by nccchirag** [reply]
Thank you so much for the response. will give it a shot! 

I understand the intended application is HA but I would like to use it simply for non-HA (plain mosquitto server) where in the device (tas client) can find the server and get connected without user intervention/manual configuration. Even if I make a relay board using tasmota I can use this feature to enumerate the device in my custom server. Since the feature is built-in, I wanted to give it a shot .. if it works as intended then I save the user from entering the MQTT details :)

Thank you

**Reply by sfromis** [reply]
You wrote about wanting the IP. Other boot time payloads already have the IP, along a few other fields
```
tele/xxxxxxxx/INFO1 {"Info1":{"Module":"ESP32S2","Version":"12.1.1.5(lvgl-haspmota)","FallbackTopic":"cmnd/xxxxxxxx_fb/","GroupTopic":"cmnd/tasmotas/"}}
tele/xxxxxxxx/INFO2 {"Info2":{"WebServerMode":"Admin","Hostname":"xxxxxxxx","IPAddress":"192.168.xx.xx"}}
tele/xxxxxxxx/INFO3 {"Info3":{"RestartReason":"Software reset digital core","BootCount":71}}
```
This thread does explain most of the Tasmota discovery payload fields. https://github.com/arendst/Tasmota/issues/9267

**Reply by nccchirag** [reply]
true, but those are only after the connection is established to the MQTT server. By IP I meant the mDNS service which finds the server automatically .. maybe I was not able to explain the application accurately 🙏

As per both your suggestions I compiled the build successfully including PING! Now the tas-client finds the server & connects to it automatically on boot 😊

Thank you again for your support! 🤝🏻

**Comment by s-hadinger** [comment, marked as answer]
That's an old exclusion we had to set up on ESP8266 because both features would not fit at the same time due to flash constraints. It should be less of an issue now.

**Reply by nccchirag** [reply]
Thank you so much for the response, I will try compiling the custom build by removing the exclusion condition. However, would you recommend wrapping it around #ifdef ESP8266?

Thank you
